src/terms/Hamiltonian.jl

```diff
@@ -0,0 +1,96 @@
+### A Hamiltonian is composed of blocks (kpoints), which have a list of RFO corresponding to each term
+# This is the "high-level" interface, provided for convenience
+using Printf
+
+struct HamiltonianBlock
+    basis::PlaneWaveBasis
+    kpoint::Kpoint
+    # The Hams are vectors of RealFourierOperator, not typed because of type invariance issues.
+    Hams::Vector # the original list of RFOs (as many as there are terms), kept for easier exploration
+    optimized_Hams::Vector # the optimized list of RFOs, to be used for applying
+    HamiltonianBlock(basis, kpt, Hams) = new(basis, kpt, Hams, optimize_RFOs_(Hams))
+end
+struct Hamiltonian
+    basis::PlaneWaveBasis
+    Hks::Vector{HamiltonianBlock}
+end
+
+# Loop through bands, IFFT to get ψ in real space, loop through terms, FFT and accumulate into Hψ
+# Do this band by band to conserve memory
+# As an optimization we special-case nonlocal operators to apply them
+# instead on the full block and benefit from BLAS3
+@views function LinearAlgebra.mul!(Hψ::AbstractArray, H::HamiltonianBlock, ψ::AbstractArray)
```

Comment:
Wow ... this routine turned out to read really nicely albeit it is pretty optimised in what it does! Great job!